wip: automatic docs generation

jlipps opened the pull request:
Figured this PR could be a basis for collaboration on automatic docs generation for drivers and plugins.

In its initial state, it shows examples of minimal updates to docstrings that could be used in conjunction with a custom docs generation plugin (for e.g. typedoc or jsdoc).

To take one example:

```js
/**
 * Get the text of an alert
 *
 * @appiumCommand
 * @returns {Promise<string>}
 */
commands.getAlertText = async function getAlertText() {
  this.assertAlert();
  return this.appModel.alertText();
};
```

All we have added to the docstring here is the "tag" `@appiumCommand`. This is what I see happening as a result in the plugin:

0. The plugin determines the function name for all functions tagged with `@appiumCommand`
1. The plugin looks up information about this command in 3 places:
    - Appium's route map
    - The extension's `newMethodMap`
    - The extension's `executeMethodMap`
2. Using the information found, it constructs appropriate documentation output  (e.g., command name, command parameters and types, response values, and W3C protocol information like HTTP method and route. Response values should be adjusted from e.g., `Promise<null>` to `null`, since while `Promise<null>` is the accurate type in our code, it probably doesn't map to anything for an end user, who may be using Java or Python or Ruby.
3. Documentation should be organized so the user knows how to use these various commands. E.g. separated into sections like "Standard appium commands", "Custom appium commands", and "Execute method commands", since the way of accessing these command types might differ.

Note there is a 4th category of commands we want to document, which have no examples in the fake-driver project. These are proxied commands. For example, the Android driver, when in proxy mode, supports the `getUrl` command. But this function is not represented anywhere in the driver codebase, because it is only proxied to Chromedriver. We might still want to allow driver authors to note that these commands are supported (rather than a hand-wavy "we support whatever chromedriver supports when in proxy mode"). So we could allow "empty" docstrings that look like this, maybe at the end of a file:

```js
/**
 * @appiumProxyCommand {Chromedriver} getUrl
 */
```

(TBD whether these should note param/return info; theoretically we can back these out of an appium routes lookup for any proxying to a standard webdriver endpoint)

In terms of the overall architecture of this system, I take it the following should also be requirements:
- docs generation system abstracted and made part of generic tooling that 3rd parties can run in their own extensions. (e.g., a 3rd party plugin author can `npm i @appium/docs && appium-docs gen-commands` and see some pretty output somewhere in HTML or Markdown (preferably markdown for easy copy/paste into READMEs)). probably should support a variety of configuration options that will become clear
- we use this system to generate docs for our own drivers/plugins
- some kind of integration should be possible with our own broader docs generation system. if we are using mkdocs for our drivers and plugins, there should be a way for us to pull in and link these command docs

Still a lot that can be discussed and figured out, for example do we have one doc page per command? Or put them all in a big table? Or different modes?

boneskull committed:
feat(): experimental support for typedoc generation

Run `npm run typedoc` to generate docs in `typedoc-docs` dir.

I thought I'd take a few mins and set this up.  It was _mostly_ straightforward, except:

1. Figuring out the "entry point" for each package was difficult until I realized `.js` entry points are assumed to be compiled, and thus require a `sourceMappingUrl` reference. IMO this is probably a bug in TypeDoc
2. The `sourceMappingUrl` must reference an external sourcemap file; not an inline source map.  If it's an inline source map, the system tries to `fs.open` a large data-url.
3. The babel config does not control whether external sourcemaps are created; `@babel/cli` does. So I needed to change those.
4. Practically speaking, I don't see any drawback to external sourcemaps _yet_
5. TypeDoc has poor support for cross-package type references in monorepos, and I had to add a plugin which implements a workaround
6. Per some suggestions in the docs, I modified the order of references in the root `tsconfig.json` to start with "the biggest one first"

I didn't modify any types or docstrings anywhere.

# Conflicts:
#	packages/base-plugin/package.json
#	tsconfig.json

jlipps committed:
wip: add some demos of docstrings to trigger auto command docs gen

boneskull commented:
does it make more sense for the driver author to specify what _doesn't_ proxy to chromedriver?

jlipps commented:
> does it make more sense for the driver author to specify what doesn't proxy to chromedriver?

I think one option for driver authors is to just have something in their docs that says "when in a webview context, commands are generally proxied to chromedriver. see chromedriver / w3c docs for more information about what commands are available."

the option I'm proposing here is for driver authors that want to have a more self-contained set of docs for their users.

at this point specifying IN vs OUT is kinda arbitrary, but I'd say that specifying OUT would be a moving target. IMO it'd be better to specify IN since theoretically proxied drivers like chromedriver are going to be a smaller set of commands?

jlipps commented:
i'm also fine if we decide that `@appiumProxyCommand` is not on the table for this iteration of docs preparedness for appium 2. we can have the proxy mode copout message on all our docs.

boneskull committed:
chore(base-driver,base-plugin,execute-driver-plugin,fake-driver,fake-plugin,images-plugin): update some types for typedoc

boneskull committed:
chore(types): address some shortcomings of method maps and constraints

This change:

1. Makes both `MethodMap` and `ExecuteMethodMap` considered "read only".  This is not _entirely_ true, as there's a bit of code that munges `ExecuteMethodDef` objects, but it's good enough for usage.
2. They both now accept a generic `D`, which is by default `Driver`.  Since a `Plugin` can also use `newMethodMap`, there's a conditional type here to add everything from `ExternalDriver` only if it's a `Driver`.  This generic allows proper validation of the map itself (command names must exist!)
3. `Constraints` is now also considered `Readonly`, since it is, in practice.
4. `executeMethodMap` is now actually optional
5. Renamed `Method` to `MethodDef` for consistency

looks like some formatting happened too

boneskull committed:
feat(typedoc-plugin-appium): typedoc plugin larvae

The first bit of a typedoc plugin for Appium to support `@appiumCommand`, etc.

There was some sort of misunderstanding between Babel and TypeScript which resulted in parts of TypeDoc's API being unavailable.  I did not want to spend a lot of time trying to figure it out, so I just wrote the plugin in TypeScript, since that seems to work.

Right now it's able to grok `executeMethodMap`, `newMethodMap`, and the bigass `METHOD_MAP` in `@appium/base-driver`.  It does not yet deal with tags nor does it generate any documentation, and in fact the current main feature is that it causes `typedoc` to exit before generating any documentation.

boneskull committed:
chore(driver-test-support,plugin-test-support): generate external source maps in build

boneskull committed:
chore(): update vscode debug profiles
